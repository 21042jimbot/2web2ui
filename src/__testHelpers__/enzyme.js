export const findButton = (wrapper, buttonLabel) =>
  wrapper.findWhere(
    (node) => node.type() === 'button' && new RegExp(buttonLabel).test(node.text())
  );
