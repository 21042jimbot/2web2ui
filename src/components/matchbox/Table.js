import React from 'react';
import { Table as OGTable } from '@sparkpost/matchbox';
import { Table as HibanaTable } from '@sparkpost/matchbox-hibana';
import { useHibana } from 'src/context/HibanaContext';
import { omitSystemProps } from 'src/helpers/hibana';

HibanaTable.displayName = 'HibanaTable';
HibanaTable.Cell.displayName = 'HibanaTableCell';
HibanaTable.HeaderCell.displayName = 'HibanaTableHeaderCell';
HibanaTable.Row.displayName = 'HibanaTableRow';
OGTable.displayName = 'OGTable';
OGTable.Cell.displayName = 'OGTableCell';
OGTable.HeaderCell.displayName = 'OGTableHeaderCell';
OGTable.Row.displayName = 'OGTableRow';

export function Table(props) {
  const [state] = useHibana();
  const { isHibanaEnabled } = state;

  if (!isHibanaEnabled) {
    return <OGTable {...omitSystemProps(props)} />;
  }
  return <HibanaTable {...props} />;
}

export function Cell(props) {
  const [state] = useHibana();
  const { isHibanaEnabled } = state;

  if (!isHibanaEnabled) {
    return <OGTable.Cell {...omitSystemProps(props)} />;
  }
  return <HibanaTable.Cell {...props} />;
}

export function HeaderCell(props) {
  const [state] = useHibana();
  const { isHibanaEnabled } = state;

  if (!isHibanaEnabled) {
    return <OGTable.HeaderCell {...omitSystemProps(props)} />;
  }
  return <HibanaTable.HeaderCell {...props} />;
}

export function Row(props) {
  const [state] = useHibana();
  const { isHibanaEnabled } = state;

  if (!isHibanaEnabled) {
    return <OGTable.Row {...omitSystemProps(props)} />;
  }
  return <HibanaTable.Row {...props} />;
}

Cell.displayName = 'Table.Cell';
HeaderCell.displayName = 'Table.HeaderCell';
Row.displayName = 'Table.Row';

Table.Cell = Cell;
Table.HeaderCell = HeaderCell;
Table.Row = Row;

export default Table;
