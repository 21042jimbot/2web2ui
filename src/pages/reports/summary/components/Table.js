import React, { Component } from 'react';
import { connect } from 'react-redux';
import cx from 'classnames';

import { getTableData } from 'src/actions/summaryChart';
import { addFilters } from 'src/actions/reportFilters';
import typeaheadCacheSelector from 'src/selectors/reportFilterTypeaheadCache';
import { hasSubaccounts } from 'src/selectors/subaccounts';

import { TableCollection, Unit, Loading } from 'src/components';
import { Empty } from 'src/components';
import { Panel, Select, Grid, UnstyledLink } from '@sparkpost/matchbox';
import { GROUP_CONFIG } from './tableConfig';
import _ from 'lodash';

import styles from './Table.module.scss';

export class Table extends Component {
  handleGroupChange = (e) => {
    this.props.getTableData({ groupBy: e.target.value });
  }

  handleRowClick = (item) => {
    this.props.addFilters([item]);
    this.props.refresh();
  }

  getColumnHeaders() {
    const { metrics, groupBy } = this.props;

    const primaryCol = {
      label: GROUP_CONFIG[groupBy].label,
      className: styles.HeaderCell
    };

    const metricCols = metrics.map(({ label, key }) => ({
      key,
      label: <div className={styles.RightAlign}>{ label }</div>,
      className: cx(styles.HeaderCell, styles.NumericalHeader)
    }));

    return [primaryCol, ...metricCols];
  }

  getRowData = () => {
    const { metrics, groupBy, typeaheadCache } = this.props;
    const group = GROUP_CONFIG[groupBy];

    return (row) => {
      let value = row[group.keyName];
      let filter = {
        type: group.label,
        value
      };

      // Matches typeahead filter object for subaccounts
      if (filter.type === 'Subaccount') {
        const subaccount = _.find(typeaheadCache, { type: 'Subaccount', id: filter.value });
        const id = filter.value;
        value = _.get(subaccount, 'value') || `Deleted (ID ${filter.value})`;

        if (filter.value === 0) {
          value = 'Master Account (ID 0)';
        }

        filter = { ...filter, value, id };
      }

      const primaryCol = <UnstyledLink onClick={() => this.handleRowClick(filter)}>{ value }</UnstyledLink>;

      const metricCols = metrics.map((metric) => (
        <div className={styles.RightAlign}>
          <Unit value={row[metric.key]} unit={metric.unit}/>
        </div>
      ));

      return [primaryCol, ...metricCols];
    };
  }

  getSelectOptions = () => {
    const options = _.keys(GROUP_CONFIG).map((key) => ({ value: key, label: GROUP_CONFIG[key].label }));

    if (!this.props.hasSubaccounts) {
      _.remove(options, { value: 'subaccount' });
    }

    return options;
  }

  renderTable() {
    const { tableData, tableLoading, groupBy } = this.props;
    const rowKeyName = GROUP_CONFIG[groupBy].keyName;

    if (tableLoading) {
      return (
        <div className={styles.LoadingSection}>
          <div className={styles.Loading}><Loading /></div>
        </div>
      );
    }

    if (!tableData.length) {
      return <Empty message='There is no data to display' />;
    }

    return (
      <TableCollection
        rowKeyName={rowKeyName}
        columns={this.getColumnHeaders()}
        getRowData={this.getRowData()}
        pagination
        defaultPerPage={10}
        rows={tableData}
        filterBox={{ show: false }}
      />
    );
  }

  render() {
    return (
      <Panel>
        <Panel.Section>
          <Grid>
            <Grid.Column xs={12} md={5} lg={4}>
              <Select
                label='Group By'
                options={this.getSelectOptions()}
                value={this.props.groupBy}
                disabled={this.props.tableLoading}
                onChange={this.handleGroupChange}/>
            </Grid.Column>
            <Grid.Column></Grid.Column>
          </Grid>
        </Panel.Section>
        { this.renderTable() }
      </Panel>
    );
  }
}

const mapStateToProps = (state) => ({
  typeaheadCache: typeaheadCacheSelector(state),
  hasSubaccounts: hasSubaccounts(state),
  ...state.summaryChart
});
export default connect(mapStateToProps, { getTableData, addFilters })(Table);
