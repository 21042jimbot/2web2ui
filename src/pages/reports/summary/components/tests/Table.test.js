import React from 'react';
import { Table } from '../Table';
import { shallow, mount } from 'enzyme';
import _ from 'lodash';

describe('Summary Table', () => {

  let wrapper;

  const props = {
    addFilters: jest.fn(),
    _getTableData: jest.fn(),
    refresh: jest.fn(),
    metrics: [
      { key: 'metric_1' },
      { key: 'metric_2', unit: 'millisecond' }
    ],
    groupBy: 'domain',
    typeaheadCache: [
      { type: 'Subaccount', id: 555, value: 'sub 1 name' }
    ],
    tableData: [],
    tableLoading: false,
    hasSubaccounts: false,
    searchOptions: {
      filters: []
    }
  };

  const data = [
    { subaccount_id: 0, metric_1: 123, metric_2: 12345 },
    { subaccount_id: 555, metric_1: 123, metric_2: 12345 },
    { subaccount_id: 1010, metric_1: 123, metric_2: 12345 }
  ];

  beforeEach(() => {
    wrapper = shallow(<Table {...props} />);
  });

  it('should render with no data', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with data', () => {
    wrapper.setProps({ tableData: data });
    expect(wrapper.find('TableCollection')).toMatchSnapshot();
  });

  it('should render loading', () => {
    wrapper.setProps({ tableLoading: true });
    expect(wrapper.find('Loading')).toHaveLength(1);
  });

  describe('group by', () => {
    let snaps;
    beforeEach(() => {
      wrapper.setProps({ groupBy: 'subaccount' });
      const func = wrapper.instance().getRowData();
      const results = _.flatten(data.map(func));

      snaps = _.map(results, (item) => mount(item));
    });

    it('should render row data', () => {
      expect(snaps).toMatchSnapshot();
    });

    it('should handle row click', () => {
      snaps[0].find('UnstyledLink').simulate('click');
      expect(props.addFilters).toHaveBeenCalledWith([{ id: 0, type: 'Subaccount', value: 'Master Account (ID 0)' }]);
    });
  });

  it('should render with aggregate data', () => {
    const aggData = [
      { metric_1: 987, metric_2: 654, metric_3: 321 }
    ];
    const metrics = [
      { key: 'metric_1' },
      { key: 'metric_2', unit: 'millisecond' },
      { key: 'metric_3' }
    ];

    wrapper.setProps({ metrics, groupBy: 'aggregate', tableData: aggData });
    expect(wrapper.find('TableCollection')).toMatchSnapshot();
  });
});
