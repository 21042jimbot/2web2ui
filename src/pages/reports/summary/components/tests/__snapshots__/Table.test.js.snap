// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary Table  should render columns 1`] = `
Array [
  Object {
    "className": undefined,
    "label": "Recipient Domain",
  },
  Object {
    "className": "",
    "key": "metric_1",
    "label": <div
      className={undefined}
    />,
  },
  Object {
    "className": "",
    "key": "metric_2",
    "label": <div
      className={undefined}
    />,
  },
]
`;

exports[`Summary Table  should render row data & handle click 1`] = `
Array [
  <UnstyledLink
    onClick={[Function]}
  >
    <a
      onClick={[Function]}
    >
      Master Account (ID 0)
    </a>
  </UnstyledLink>,
  <div>
    <Unit
      value={123}
    >
      123
    </Unit>
  </div>,
  <div>
    <Unit
      unit="millisecond"
      value={12345}
    >
      12,345
    </Unit>
  </div>,
  <UnstyledLink
    onClick={[Function]}
  >
    <a
      onClick={[Function]}
    >
      sub 1 name
    </a>
  </UnstyledLink>,
  <div>
    <Unit
      value={123}
    >
      123
    </Unit>
  </div>,
  <div>
    <Unit
      unit="millisecond"
      value={12345}
    >
      12,345
    </Unit>
  </div>,
  <UnstyledLink
    onClick={[Function]}
  >
    <a
      onClick={[Function]}
    >
      Deleted (ID 1010)
    </a>
  </UnstyledLink>,
  <div>
    <Unit
      value={123}
    >
      123
    </Unit>
  </div>,
  <div>
    <Unit
      unit="millisecond"
      value={12345}
    >
      12,345
    </Unit>
  </div>,
]
`;

exports[`Summary Table  should render subaccount option 1`] = `
<Select
  disabled={false}
  label="Group By"
  onChange={[Function]}
  options={
    Array [
      Object {
        "label": "Recipient Domain",
        "value": "domain",
      },
      Object {
        "label": "Sending Domain",
        "value": "sending-domain",
      },
      Object {
        "label": "Campaign",
        "value": "campaign",
      },
      Object {
        "label": "Template",
        "value": "template",
      },
      Object {
        "label": "Subaccount",
        "value": "subaccount",
      },
      Object {
        "label": "Sending IP",
        "value": "sending-ip",
      },
      Object {
        "label": "IP Pool",
        "value": "ip-pool",
      },
    ]
  }
  placeholderValue=""
  value="domain"
/>
`;

exports[`Summary Table  should render with no data 1`] = `
<Panel>
  <Panel.Section>
    <Grid>
      <Grid.Column
        lg={4}
        md={5}
        xs={12}
      >
        <Select
          disabled={false}
          label="Group By"
          onChange={[Function]}
          options={
            Array [
              Object {
                "label": "Recipient Domain",
                "value": "domain",
              },
              Object {
                "label": "Sending Domain",
                "value": "sending-domain",
              },
              Object {
                "label": "Campaign",
                "value": "campaign",
              },
              Object {
                "label": "Template",
                "value": "template",
              },
              Object {
                "label": "Sending IP",
                "value": "sending-ip",
              },
              Object {
                "label": "IP Pool",
                "value": "ip-pool",
              },
            ]
          }
          placeholderValue=""
          value="domain"
        />
      </Grid.Column>
      <Grid.Column />
    </Grid>
  </Panel.Section>
  <Empty
    message="There is no data to display"
  />
</Panel>
`;
